src/uct/gaudi/base/gaudi_topo.c

```diff
@@ -0,0 +1,1929 @@
+    unsigned conn_idx;
+    unsigned i, j;
+    unsigned max_num_connections;
+    uint16_t hnic_vendor_id;
+
+    max_num_connections            = uct_gaudi_topo_ctx.num_gaudi_devices *
+                                     uct_gaudi_topo_ctx.num_hnic_devices;
+    uct_gaudi_topo_ctx.connections = ucs_calloc(max_num_connections,
+                                                sizeof(uct_gaudi_connection_t),
+                                                "gaudi_connections");
+    if (!uct_gaudi_topo_ctx.connections) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    conn_idx = 0;
+    for (i = 0; i < uct_gaudi_topo_ctx.num_gaudi_devices; i++) {
+        gaudi      = uct_gaudi_topo_ctx.gaudi_devices[i];
+        gaudi_name = uct_gaudi_topo_ctx.gaudi_devices_names[i];
+        numa1      = ucs_topo_sys_device_get_numa_node(gaudi);
+
+        for (j = 0; j < uct_gaudi_topo_ctx.num_hnic_devices; j++) {
+            hnic           = uct_gaudi_topo_ctx.hnic_devices[j];
+            hnic_vendor_id = uct_gaudi_topo_ctx.hnic_vendor_ids[j];
+            if (!uct_gaudi_is_hnic_active(hnic, hnic_vendor_id)) {
+                continue;
+            }
+            numa2 = ucs_topo_sys_device_get_numa_node(hnic);
+
+            conn               = &uct_gaudi_topo_ctx.connections[conn_idx];
+            conn->gaudi_device = gaudi;
+            ucs_strncpy_safe(conn->gaudi_dev_name, gaudi_name,
+                             GAUDI_DEVICE_NAME_LEN);
+            conn->hnic_device      = hnic;
+            conn->hnic_vendor_id   = hnic_vendor_id;
+            conn->common_numa_node = ((numa1 == numa2) &&
+                                      (numa1 != UCS_NUMA_NODE_UNDEFINED)) ?
+                                             numa1 :
+                                             UCS_NUMA_NODE_UNDEFINED;
+
+            status = uct_gaudi_estimate_distance(gaudi, hnic, &conn->distance);
+            if (status != UCS_OK) {
+                ucs_debug("failed to estimate distance between gaudi %u and "
+                          "hnic %u",
+                          gaudi, hnic);
+                conn->distance = ucs_topo_default_distance;
+            }
+            conn_idx++;
+        }
+    }
+
+    uct_gaudi_topo_ctx.num_connections = conn_idx;
+    return UCS_OK;
+}
+
+/* Fetch precomputed distance for (gaudi, hnic) from the connections list */
+static int uct_gaudi_lookup_distance(ucs_sys_device_t gaudi_device,
+                                     ucs_sys_device_t hnic_device,
+                                     ucs_sys_dev_distance_t *distance)
+{
+    const uct_gaudi_connection_t *conn;
+    unsigned i;
+
+    /* Lookup connection matrix for distance metric */
+    for (i = 0; i < uct_gaudi_topo_ctx.num_connections; i++) {
+        conn = &uct_gaudi_topo_ctx.connections[i];
+        if (conn->gaudi_device == gaudi_device &&
+            conn->hnic_device == hnic_device) {
+            if (distance) {
+                *distance = conn->distance;
```

Comment:
looks like it can not be called without valid distance pointer
i'd remove this check and add assert instead (like `ucs_assert(distance != NULL)`)